File: c/ch1/e.c
#include <petsc.h>

int main(int argc, char **argv) {
  PetscMPIInt    rank;
  PetscInt       i;
  PetscReal      localval, globalsum;

  PetscCall(PetscInitialize(&argc,&argv,NULL,
      "Compute e in parallel with PETSc.\n\n"));
  PetscCall(MPI_Comm_rank(PETSC_COMM_WORLD,&rank));

  // compute  1/n!  where n = (rank of process) + 1
  localval = 1.0;
  for (i = 2; i < rank+1; i++)
      localval /= i;

  // sum the contributions over all processes
  PetscCall(MPI_Allreduce(&localval,&globalsum,1,MPIU_REAL,MPIU_SUM,
      PETSC_COMM_WORLD));

  // output estimate of e and report on work from each process
  PetscCall(PetscPrintf(PETSC_COMM_WORLD,
      "e is about %17.15f\n",globalsum));
  PetscCall(PetscPrintf(PETSC_COMM_SELF,
      "rank %d did %d flops\n",rank,(rank > 0) ? rank-1 : 0));
  PetscCall(PetscFinalize());
  return 0;
}

File: python/ch1/e.py
#!/usr/bin/env python3
'''Compute e in parallel with PETSc.'''

import sys
import petsc4py
from petsc4py import PETSc
from mpi4py import MPI
petsc4py.init(sys.argv)

comm = PETSc.COMM_WORLD.tompi4py()
rank = comm.Get_rank()

# compute  1/n!  where n = (rank of process) + 1
localval = 1.0
for i in range(2,rank+1):
    localval = localval / i

# sum the contributions over all processes
globalsum = comm.allreduce(localval, op=MPI.SUM)

# output estimate of e and report on work from each process
PETSc.Sys.Print('e is about %17.15f' % globalsum, comm=PETSc.COMM_WORLD)
if rank > 0:
    q = rank-1
else:
    q = 0
PETSc.Sys.Print('rank %d did %d flops' % (rank,q), comm=PETSc.COMM_SELF)
